From 304b58f372256629ef725e08001a910ffa6d8b14 Mon Sep 17 00:00:00 2001
From: "Zheng, Xiao" <xiao.zheng@intel.com>
Date: Mon, 20 Jul 2015 19:07:34 +0800
Subject: [PATCH 130/403] i915: wait_for_atomic will spin on udelay instead of
 jiffies

The fix is to enhance stability test [MTBF]
Current wait_for_atomic in i915 driver side will spin on jiffies, however this
function was used in irq disabled environment and sometimes cause deadlock
since timer are disabled and jiffies will not be updated.

Signed-off-by: Zheng, Xiao <xiao.zheng@intel.com>
---
 drivers/gpu/drm/i915/intel_drv.h |   20 +++++++++++++++++---
 1 file changed, 17 insertions(+), 3 deletions(-)

diff --git a/drivers/gpu/drm/i915/intel_drv.h b/drivers/gpu/drm/i915/intel_drv.h
index 1a3bbdc..f9aeb80 100644
--- a/drivers/gpu/drm/i915/intel_drv.h
+++ b/drivers/gpu/drm/i915/intel_drv.h
@@ -63,10 +63,24 @@
 	ret__;								\
 })
 
+/* invoked likely in irq disabled condition */
+#define _wait_for2(COND, MS) ({						\
+	unsigned long cnt = MS*100;					\
+	int ret__ = 0;							\
+	while (!(COND)) {						\
+		if (!(--cnt)) {						\
+			ret__ = -ETIMEDOUT;				\
+			break;						\
+		}							\
+		udelay(10);						\
+	}								\
+	ret__;								\
+})
+
 #define wait_for(COND, MS) _wait_for(COND, MS, 1)
-#define wait_for_atomic(COND, MS) _wait_for(COND, MS, 0)
-#define wait_for_atomic_us(COND, US) _wait_for((COND), \
-					       DIV_ROUND_UP((US), 1000), 0)
+#define wait_for_atomic(COND, MS) _wait_for2(COND, MS)
+#define wait_for_atomic_us(COND, US) _wait_for2((COND), \
+					       DIV_ROUND_UP((US), 1000))
 
 #define KHz(x) (1000 * (x))
 #define MHz(x) KHz(1000 * (x))
-- 
1.7.10.4

